### PR TITLE
fix(database): Fix missing database logs

### PR DESCRIPTION
**Pull request by zhengkunwang223**

### CODE REVIEW

File: agent/app/service/container.go [hunk 1]
```diff
@@ -915,17 +907,10 @@ func collectLogs(params dto.StreamLog, messageChan chan<- string, errorChan chan
...
 	}
-
-	errScanner := bufio.NewScanner(stderr)
-	for errScanner.Scan() {
-		line := errScanner.Text()
-		errorChan <- fmt.Errorf("%v", line)
-	}
 	if err = dockerCmd.Wait(); err != nil {
 		errorChan <- fmt.Errorf("%v", err)
 		return
```
Comment (wanghe-fit2cloud) [inline agent/app/service/container.go:916]: The provided code has several issues and optimizations that can be made. Here are the main points:

### Issues and Problems Fixed:

1. **Duplicate `stderr` Handling**: The code tries to read from both standard output (`stdout`) and standard error (`stderr`). It attempts to create separate pipes for each but then combines them, which is unnecessary because it reads from the same place.

2. **Unnecessary Variable Initialization**: The variable `lineNumber` was declared within the loop without being used elsewhere in the function, leading to unused declaration.

3. **Error Channels not Properly Closed**: There's no explicit closing of `messageChan` or `errorChan` after use, which could lead to memory leaks if they're goroutine-safe channels.

4. **Error Handling Not Consistent**: Error handling is mostly done using `fmt.Errorf`, but there isn't consistency with how errors are returned and handled throughout the function.

### Optimizations Suggested:

1. **Single Read Pipe**: Use only one pipe to read combined standard output and error messages, reducing complexity and improving efficiency.

2. **Explicit Closing of Channels**: Add `close(messageChan)` and `close(errorChan)` statements at the end of the function to ensure resources are released correctly even if the loop exits unexpectedly.

Here's an improved version of the code incorporating these changes:

```go
func collectLogs(params dto.StreamLog, messageChan chan<- string, errorChan chan<- error) {
	cmd := exec.Command(dockerCmd.Executable(), params.Args...)
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		errorChan <- fmt.Errorf("failed to get stdout pipe: %v", err)
		return
	}

	scanner := bufio.NewScanner(stdout)

	go func() {
		defer stdout.Close()

		for scanner.Scan() {
			messageChan <- scanner.Text()
			select {
			case <-params.CancelContext.Done():
				return
			default:
			}
		}

		if err := scanner.Err(); err != nil {
			errorChan <- &dto.DrainError{
				Type:    "drain_error",
				Message: err.Error(),
			}
		}
	}()

	err = cmd.Start()
	if err != nil {
		close(messageChan)
		close(errorChan)
		errorChan <- fmt.Errorf("failed to start command: %v", err)
		return
	}

	err = cmd.Wait()
	if err != nil {
		close(messageChan)
		close(errorChan)
		errorChan <- &dto.DrainError{
			Type:    "wait_error",
			Message: err.Error(),
		}
	}
}
```

This refactored version simplifies error handling and resource management, making it more robust and easier to understand.

File: frontend/src/components/log/container/index.vue [hunk 2]
```diff
@@ -252,7 +260,7 @@ onMounted(() => {
 }
 
 .log-container {
-    height: calc(100vh - 320px);
+    height: calc(100vh - var(--custom-height, 320px));
     overflow-y: auto;
     overflow-x: auto;
     position: relative;
```
Comment (wanghe-fit2cloud) [inline frontend/src/components/log/container/index.vue:266]: The code contains a couple of improvements and corrections:

1. **Variable Naming**: The variable `highlightDiff` should probably have been renamed to something more descriptive, like `fixedHeight`, which better reflects its purpose.

2. **Style Variable**: Correcting the style for `--custom-height`. Should be `--custom-height: ${props.highlightDiff}px;`.

3. **Logical Mistake**: In the template, removing `.log-spacer` because heights were already calculated via `visibleLogs.length`.

Here is the revised code snippet with comments explaining the changes:

```diff
@@ -24,7 +24,7 @@
             {{ $t('commons.button.clean') }}
         </el-button>
     </div>
-    <div class="log-container" ref="logContainer">
+    <div class="log-container" :style=styleVarsRef value:"${props.fixedHeight}px"`
         <div class="log-spacer" :style="{ height: `${totalHeight}px` }"></div> <!-- Removed -->
         <div
             v-for="(log, index) in visibleLogs"
@@ -61,8 +61,9 @@ const props = defineProps({
         type: String,
         default: '',
     },
-    highlightDiff: {
+    fixedHeight: { // Renamed from highlightDiff to make it clearer
         type: Number,
         default: 320,
     },
 });
 
+const styleVarsRef = computed(() => ({
+    '--custom-height': `${props.fixedHeight}px`,
 }));
+
 const logVisible = ref(false);
 const logContainer = ref<HTMLElement | null>(null);
 const logs = ref<string[]>([]);
```

### Additional Suggestions

- Ensure that `totalHeight` is correctly calculated based on your requirements.
  
- Consider moving all logical calculations and constants inside methods rather than directly in props or templates if they don't change frequently. 

These adjustments might help improve maintainability and correctness of your component.

File: frontend/src/views/database/mysql/setting/index.vue [hunk 3]
```diff
@@ -86,7 +86,7 @@
                         </el-row>
                     </el-form>
                 </div>
-                <ContainerLog v-if="activeName === 'log'" :container="containerID" />
+                <ContainerLog v-if="activeName === 'log'" :container="baseInfo.containerID" :highlightDiff="350" />
                 <SlowLog
                     @loading="changeLoading"
                     @refresh="loadBaseInfo"
```
Comment (wanghe-fit2cloud) [inline frontend/src/views/database/mysql/setting/index.vue:92]: The changes suggested to the code:

1. The component `ContainerLog` is now receiving a new data property `highlightDiff`, which may be needed for highlighting specific lines in the log.
2. It's advisable to ensure that the values of `containerID` and `baseInfo.containerID` are correctly passed and used within the `<ContainerLog>` component.

These might improve maintainability or functionality depending on how the components interact with each other. Make sure that any additional logic related to this change has been adequately tested.

File: agent/app/service/container.go (hunk 1 above)
Comment (wanghe-fit2cloud) [inline agent/app/service/container.go:916]: The provided code snippet is for a function that collects both standard output (`stdout`) and standard error (`stderr`) from a Docker command in an asynchronous manner. After collecting all messages, it waits for the command to complete.

Here are some suggestions for improvement:

1. **Resource Management**:
   - Ensure that `dockerCmd` is properly closed after use. You can do this manually with `defer dockerCmd.Wait()` if necessary.

2. **Error Handling**:
   - Move the exit cases outside of the loop bodies to prevent unnecessary scanning and waiting.
     ```go
    		return
     }

     if err = scanner.Err(); err != nil {
         errorChan <- fmt.Errorf("scanner error: %v", err)
         return
     }

     // Exit after processing stderr only once (assuming no errors occurred)
     if err = dockerCmd.Wait(); err != nil && len(err.Error()) > 0 {
         errorChan <- fmt.Errorf("%v", err)
         return
     }
     ```

3. **Optimization**:
   - Combine the handling of stdout and stderr into a single scanning loop since they both have similar logic. This reduces redundancy.
     ```go
     for errScanner.Scan() {
         line := errScanner.Text()
         select {
         case messageChan <- msg(line):
             return
         case messageChan <- err(line):
             return
         default:
         }
     }
     ```
     Note that `msg` and `err` functions should be defined elsewhere to handle differentiating between stdout and stderr messages.

4. **Debugging Information**:
   - Consider logging information about what's happening during execution, especially when dealing with complex processes like container logs.

5. **Avoid Reusing Scanner Instances**:
   - Each iteration using `bufio.NewScanner` creates a new instance unnecessarily, potentially leading to resource usage inefficiencies. You can reuse a single scanner object across multiple checks within loops.
     
These changes improve the robustness and efficiency of managing the input streams and errors associated with running a Docker command asynchronously.

File: frontend/src/components/log/container/index.vue [hunk 4]
```diff
@@ -252,7 +260,7 @@ onMounted(() => {
 }
 
 .log-container {
-    height: calc(100vh - 350px);
+    height: calc(100vh - var(--custom-height, 320px));
     overflow-y: auto;
     overflow-x: auto;
     position: relative;
```
Comment (wanghe-fit2cloud) [inline frontend/src/components/log/container/index.vue:266]: The code you've provided has a few suggestions for optimization and potential improvements:

### Optimizations and Improvements:

1. **Variable Naming:** Use more descriptive variable names to improve readability.
   ```typescript
   const highlightedDiff = props.highlightDiff || 320; // Using camelCase is more common in TypeScript
   ```

2. **Computed Property Optimization:** Avoid recomputing `styleVars` every time it's used (when `highlightDiff` changes). Instead, use `watchEffect`.
   ```typescript
   let prevHighlightDiff = null;
   watchEffect(() => {
       if (prevHighlightDiff !== props.highlightDiff) {
           styleVars.value['--custom-height'] = `${props.highlightDiff}px`;
           prevHighlightDiff = props.highlightDiff;
       }
   });
   ```

3. **Avoid Redundant Style Declarations:** Directly set the height using CSS custom properties in JavaScript rather than repeating them in inline styles.

4. **Use Template Refs Properly:** Ensure that template refs are properly referenced in methods and lifecycle hooks.
   ```typescript
   onMounted(() => {
       logContainer.value?.scrollTo({ top: totalHeight, behavior: 'smooth' }); // Ensures smooth scrolling when logging starts from bottom
   });
   ```

5. **Type Annotations:** Add type annotations where necessary for better code clarity and static typing support.
   ```javascript
   const scrollLogToBottom = () => { // Adding return type for clearer function usage
       logContainer.value?.scrollTo({ top: totalHeight, behavior: 'smooth' });
   };
   ```

6. **Accessibility Considerations:** If this component logs user actions, consider adding ARIA attributes to make it accessible to screen readers.
   ```html
   <div class="log-container" :style="styleVars" role="log" aria-live="polite"> <!-- Assuming all logs are important -->
```

7. **Performance Enhancements:** Ensure other parts of the application do not interfere with performance due to excessive DOM manipulations. For example, avoid unnecessary repaints or renders during log updates.

8. **Testing:** Thoroughly test edge cases involving dynamic content and different device/viewport sizes to ensure compatibility and responsiveness.

By making these adjustments, the code will be more efficient, maintainable, and easier to understand while addressing some known issues.

File: frontend/src/views/database/mysql/setting/index.vue [hunk 5]
```diff
@@ -86,7 +86,7 @@
                         </el-row>
                     </el-form>
                 </div>
-                <ContainerLog v-if="activeName === 'log'" :container="baseInfo.containerID" />
+                <ContainerLog v-if="activeName === 'log'" :container="baseInfo.containerID" :highlightDiff="350" />
                 <SlowLog
                     @loading="changeLoading"
                     @refresh="loadBaseInfo"
```
Comment (wanghe-fit2cloud) [inline frontend/src/views/database/mysql/setting/index.vue:92]: The code you provided appears to be updating the content of an `el-row` based on the value of `activeName`. If `activeName` is equal to `'log'`, it renders either a `ContainerLog` component (if there's currently no log highlighting) or a version with `:highlightDiff="350"`.

### Potential Issues:

1. **Highlight Diff Value**: The presence of `:highlightDiff="350"` suggests that you might want to highlight changes in logs with this specific difference. It's unclear why "350" would be used here since it doesn't relate directly to how the logging should behave. Consider replacing `"350"` with something more meaningful if applicable.

2. **Error Handling**: Ensure that the `highlightDiff` prop can handle values outside its expected range smoothly without causing errors.

### Optimization Suggestions:

1. **Lazy Loading of Logs**: If loading logs takes time, consider implementing lazy loading so that only necessary logs are loaded at a time and rendered as needed.

2. **Conditional Rendering Logic**: Simplify the conditional rendering logic for better maintainability. For example:
   ```javascript
   <template>
       <div v-if="activeName === 'log'">
           <ContainerLog :container="baseInfo.containerID" v-if="!highlightedLogs"/>
           <ContainerLog :container="baseInfo.containerID" :highlightDiff="350" v-else/>
       </div>
   </template>

   <script>
   export default {
       props: {
           activeName: String,
           baseInfo: Object,
           highlightedLogs: Boolean
       }
   };
   </script>
   ```

These suggestions aim to improve the readability and performance of your code while ensuring robustness. Let me know if these points help clarify things further!
